File: cracking/arrays/Ex12.java
class Ex12 {
    static boolean isPermutation(String s1, String s2) {
        if (s1.length() != s2.length()) {
            return false;
        }
        var counters = new int[256];
        for (char c : s1.toCharArray()) {
            counters[c] += 1;
        }
        for (char c : s2.toCharArray()) {
            counters[c] -= 1;
        }
        for (int counter : counters) {
            if (counter != 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(isPermutation("a", "b"));
        System.out.println(isPermutation("abcd", "bcad"));
        System.out.println(isPermutation("c", "abc"));
    }
}

File: cracking/arrays/Ex12.py
def is_permutation(s1: str, s2: str) -> bool:
    if len(s1) != len(s2):
        return False
    counters = [0]*256
    for c in s1:
        counters[ord(c)] += 1
    for c in s2:
        counters[ord(c)] -= 1
    for counter in counters:
        if counter != 0:
            return False
    return True


assert not is_permutation("a", "b")
assert is_permutation("abcd", "bcad")
assert not is_permutation("a", "abc")